### PR TITLE
feat(lint/noSemicolonInJsx): add rule

### LINKED ISSUE #927: 📎 `lint/noSuspiciousSemicolonInJSX`: disallow suspicious `;` in JSX
### Description

I don't know if such a lint rule exists in other tools, but this is an issue I've ran into a few times before.

Basically, I'd like to have a rule to catch semicolons that may have been introduced in the JSX after a copy/paste or refactor, as these semicolons are then rendered.

Examples of invalid code:
```jsx
return (
  <div>
    <div />;
  </div>
);

return (
  <div>
    <Component>
      <div />
    </Component>;
  </div>
);
```

Examples of valid code:
```jsx
return (
  <div>
    <div />
  </div>
);

// makes it a bit more explicit that we intend to render a semicolon
return (
  <div>
    <div />
    ;
  </div>
);

// makes it a bit more explicit that we intend to render a semicolon
return (
  <div>
    <div />{';'}
  </div>
);

// use an html entity instead https://www.compart.com/en/unicode/U+003B
return (
  <div>
    <div />&#59;
  </div>
);

return (
  <div>
    {/* biome-ignore lint/??/??: rendering this semicolon is intentional */}
    <div />;
  </div>
);

return (
  <div>
    <span>;</span>
    <span />;</span />
    text; text;
    &amp;
  </div>
);

return <div />;

return <div>
  <div />
</div>;
```

It's an easy mistake to make if you don't pay attention.
Recording 2023-11-27 at 23 16 29

### PR DESCRIPTION
**Pull request by fujiyamaorange**
## Summary
Closes #927 

I referred [Contributing document](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#-contributing), but I may be missing a few things. Please let me know 🙏 

### Valid
```
const Component = () => {
    return (
        <div>
            <div />
        </div>
    );
}

const Component2 = () => {
    return (
        <div>
            <div />
            ;
        </div>
    );
}

const Component3 = () => {
    return (
        <div>
          <div />{';'}
        </div>
    );
}

const Component5 = () => {
    return (
        <div>
          <span>;</span>
          <span />;<span />
          text; text;
          &amp;
        </div>
    );
}

const Component6 = () => {
    return <div />;
}
```

### Invalid
```
const Component = () => {
    return (
        <div>
          <div />;
        </div>
    );
}

const Component2 = () => {
    return (
        <div>
          <Component>
            <div />
          </Component>;
        </div>
    );
}

const Component3 = () => (
    <div>
        <Component />;
    </div>
)
```

### CODE REVIEW

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    ///   return (
+    ///     <div>
+    ///       <span>;</span>
+    ///     </div>
+    ///   );
+    /// }
+    /// ```
+    ///
+    pub(crate) NoSemicolonInJsx {
+        version: "1.0.0",
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:46]:
```suggestion
        version: "next",
```

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let NoSemicolonInJsxState {
+            incorrect_semicolon,
+        } = state;
+
+        let diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            incorrect_semicolon.text_range(),
+            markup! {
+                "There is suspicious "<Emphasis>"Semicolon"</Emphasis>" in the JSX element."
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:87]:
```suggestion
                "There is a suspicious "<Emphasis>"semicolon"</Emphasis>" in the JSX element."
```

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            },
+        )
+        .note(markup! {
+            "Remove the "<Emphasis>"Semicolon"</Emphasis>" from the JSX element."
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:91]:
```suggestion
            "Remove the "<Emphasis>"semicolon"</Emphasis>", or move it inside a JSX element."
```

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        } = state;
+
+        let diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            incorrect_semicolon.text_range(),
+            markup! {
+                "There is suspicious "<Emphasis>"Semicolon"</Emphasis>" in the JSX element."
+            },
+        )
+        .note(markup! {
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:90]:
```suggestion
        )
        .note(markup! {
        	"This is usually the result of a typo or some refactor gone wrong."
        })
        .note(markup! {
```

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic, RuleSource, RuleSourceKind,
+};
+use biome_console::markup;
+use biome_js_syntax::{jsx_ext::AnyJsxElement, JsLanguage, JsxChildList, JsxElement};
+use biome_rowan::{AstNode, AstNodeList, SyntaxToken};
+
+declare_rule! {
+    /// Remove semicolons from JSX elements.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:9]:
```suggestion
    /// It detects possible "wrong" semicolons inside JSX elements.
    ///
    /// Semicolons that appear after a self-closing element or a closing element are usually the result of a typo of a refactor gone wrong.
```

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+pub(crate) struct NoSemicolonInJsxState {
+    incorrect_semicolon: SyntaxToken<JsLanguage>,
+}
+
+impl Rule for NoSemicolonInJsx {
+    type Query = Ast<AnyJsxElement>;
+    type State = NoSemicolonInJsxState;
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:60]: You can remove `NoSemicolonInJsxState` and just use `TextRange` as state

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let NoSemicolonInJsxState {
+            incorrect_semicolon,
+        } = state;
+
+        let diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            incorrect_semicolon.text_range(),
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:85]: You should use `text_trimmed_range()`. The API `text_range` includes all trivia (whitespace, newlines, etc.)

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        });
+        Some(diagnostic)
+    }
+}
+
+fn has_suspicious_semicolon(node: &JsxChildList) -> Option<SyntaxToken<JsLanguage>> {
+    node.iter().find_map(|c| {
+        let jsx_text = c.as_jsx_text()?;
+        let jsx_text_value = jsx_text.value_token().ok()?;
+        if jsx_text_value.text().starts_with(";\n") {
```
Comment (ematipico) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:101]: We should also check for `\r` and `\r\n`

**Review by ematipico — COMMENTED** [review]
Thank you! I applied some suggestions for the documentation. Feel free to apply them. You will have to run `cargo t` and `just gen-lint` again

**Comment by fujiyamaorange** [comment]
@ematipico 
Thank you very much for taking the time to review and for offering your valuable insights.

I will attend to all the suggested modifications. May I kindly request a bit of your patience during this process? 🙏

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
+use biome_analyze::{
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:1]: The rule is unstable, thus it should be in the nursery folder.

File: crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    ///     </div>
+    ///   );
+    /// }
+    /// ```
+    ///
+    pub(crate) NoSemicolonInJsx {
+        version: "next",
+        name: "noSemicolonInJsx",
+        source: RuleSource::Eslint("no-semicolons-in-jsx"),
+        source_kind: RuleSourceKind::Inspired,
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/semantic_analyzers/suspicious/no_semicolon_in_jsx.rs:51]: This rule doesn't exist in ESLint, we should remove `source` and `source_kind`.
```suggestion
```

File: crates/biome_diagnostics_categories/src/categories.rs [hunk 11]
```diff
@@ -212,6 +213,7 @@ define_categories! {
     "lint/suspicious/noSelfCompare": "https://biomejs.dev/linter/rules/no-self-compare",
     "lint/suspicious/noShadowRestrictedNames": "https://biomejs.dev/linter/rules/no-shadow-restricted-names",
     "lint/suspicious/noSparseArray": "https://biomejs.dev/linter/rules/no-sparse-array",
+    "lint/suspicious/noSemicolonInJsx": "https://biomejs.dev/linter/rules/no-semicolons-in-jsx",
```
Comment (ematipico) [inline crates/biome_diagnostics_categories/src/categories.rs:216]:
```suggestion
```

**Review by ematipico — APPROVED** [review]

**Review by Conaclos — APPROVED** [review]
Great job!

**Comment by fujiyamaorange** [comment]
@ematipico @Conaclos 
May I merge this?? 💭

**Pull request merged**